src/transformers/testing_utils.py

```diff
@@ -2096,3 +2117,84 @@ def _find(self, tests, obj, name, module, source_lines, globs, seen) -> None:
         for test in finder.find(module, module.__name__):
             if test.examples:  # skip empty doctests and cuda
                 yield DoctestItem.from_parent(self, name=test.name, runner=runner, dtest=test)
+
+
+def _device_agnostic_dispatch(device: str, dispatch_table: Dict[str, Callable], *args, **kwargs):
+    if device not in dispatch_table:
+        return dispatch_table["default"](*args, **kwargs)
+
+    fn = dispatch_table[device]
+
+    # Some device agnostic functions return values. Need to guard against `None`
+    # instead at user level.
+    if fn is None:
+        return None
+    return fn(*args, **kwargs)
+
+
+# Mappings from device names to callable functions to support device agnostic
+# testing.
+ACCELERATOR_MANUAL_SEED = {"cuda": torch.cuda.manual_seed, "cpu": None, "default": torch.manual_seed}
+ACCELERATOR_EMPTY_CACHE = {"cuda": torch.cuda.empty_cache, "cpu": None, "default": None}
+ACCELERATOR_DEVICE_COUNT = {"cuda": torch.cuda.device_count, "cpu": lambda: 0, "default": lambda: 1}
```

Comment:
We need to guard this inside `is_torch_availabe()`. If not, set all of them to be empty dict

Reply:
might be better to guard the whole block? starting from line 2154?

Reply:
not sure I understand. You mean only guard the lines after 2154?

Reply:
Yep, guard from this line
https://github.com/huggingface/transformers/blob/c961bd633e29b98e598011643de4e166fb136dd6/src/transformers/testing_utils.py#L2154

None of the logic is necessary outside of a torch context.

Reply:
The definition

> ACCELERATOR_MANUAL_SEED = {"cuda": torch.cuda.manual_seed, "cpu": torch.manual_seed, "default": torch.manual_seed}

etc. already needs to be guarded, as it uses `torch`.

The definition of `accelerator_manual_seed` etc. are imported in other files, so they can't be guarded here, otherwise import will fail.

The block in `if "TRANSFORMERS_TEST_DEVICE_SPEC" in os.environ` could be guarded.

Reply:
Ah sorry, misunderstood the line you meant.

Will apply these changes~